core/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -301,9 +337,37 @@ private boolean assertSerialization(DocumentMapper mapper) {
         return true;
     }
 
-    protected void checkNewMappersCompatibility(Collection<ObjectMapper> newObjectMappers, Collection<FieldMapper> newFieldMappers, boolean updateAllTypes) {
+    /**
+     * Check that fields are defined only once. It is possible to define fields
+     * several times eg. if a field is defined once via 'fields' in the mapping
+     * and once from a field mapper.
+     */
+    private void checkUniqueness(Collection<ObjectMapper> objectMappers, Collection<FieldMapper> fieldMappers) {
```

Comment:
This would help for 3 cases:
 - a field name that collides with a metadata mapper (eg. by using a field called _id)
 - collisions between multifields that are defined in the mapping and in a fieldmapper itself (see changes to ExternalMapperTests)
 - collisions between object field and actual fields